File: Residential_Controller.cs
using System;
using System.Collections.Generic;
namespace myApp
{
    class Column
    {
        public int nbFloor;
        public int nbElevators;
        public List<Elevator> elevatorsList; 
        public Column (int nbFloor, int nbElevators){
            this.nbFloor = nbFloor;
            this.nbElevators = nbElevators;
            this.elevatorsList = new List<Elevator> ();
            for (int i =0; i < this.nbElevators; i++) {
              elevatorsList.push(new Elevator(i + 1, this.nbFloor));
          
            
        }
    }
    class Button
    {
        public string direction;
        public int requestFloor;

        public Button (string direction, int requestFloor)
        {
           this.direction = direction;
           this.requestFloor = requestFloor;
           this.activateButton = false;
        } 
    }


    class InsideButton 
    {
       public int floor;
       public string status; 
       public InsideButton (int floor) 
       {
           this.floor = floor;
           this.status = 'desactivated';
       }
    }

    class ElevatorController
    {
       public int nbFloor;
       public List<button> buttonList; 
       public int nbElevators;
       public int selectedElevator;

       public ElevatorController(int nbFloor, int nbElevators)
       {
           this.nbFloor = nbFloor;
           this.nbElevators = nbElevators;
           this.column = new Column (int nbFloor, int nbElevators);
           this.buttonList = new List<button> ();
           //List<Button>.push(new  Button());
       }
       public void RequestElevator(int FloorNumber, string Direction)
       {
           int selectedElevator = this.FindElevator(int FloorNumber, string direction);
           selectedElevator.addFloorToList(int FloorNumber);
           selectedElevator.activateInsideButton(int FloorNumber);
           Console.WriteLine("Request Elevator on floor " + FloorNumber.ToString() + ", going " + Direction);
       }
       public void RequestFloor(string Elevator, string Direction)
       {
           Elevator.activateInsideButton(int FloorNumber);
           Elevator.addFloorToList(int FloorNumber);
           Elevator.moveNext();
           Console.WriteLine("Request Floor number " + FloorNumber.ToString() + ", going " + Direction);
       }

       public void FindElevator(int FloorNumber)
       {
            int distanceFloor = 999;
            int selectedElevator = null;
            for(int i = 0; i< this.column.elevatorsList.Count(); i++){
            int differenceFloor = Math.abs(FloorNumber - this.column.elevatorsList[i].currentFloor);

            if (differenceFloor < distanceFloor){
                distanceFloor = differenceFloor;
                selectedElevator = this.column.elevatorList[i];
            
            }
            return selectedElevator;
            }      
           Console.WriteLine("FindElevator " + FloorNumber.ToString());
       
        }

    
    class Elevator
    {
       public int elevatorNumber;
       public int FloorNumber; 
       public int nbFloor;
       public string status;
       public List<floor> floorList; 
       public int currentFloor; 
       public Elevator(int elevatorNumber, int nbFloor)
       {
           this.elevatorNumber = elevatorNumber; 
           this.direction = "NONE";
           this.status = "idle";
           this.floorList = new List<floor> ();
           this.buttonList = new List<Button> ();

        for ( int i =0; i < nbFloor; i++)this.elevatorNumber = elevatorNumber; 
            this.buttonList = buttonList.push(new InsideButton(i));
        
            return currentFloor = 1; 
        }
       


       public void moveNext()
       {
            var FloorNumber = floorList.Pop();
            
            if (this.currentFloor > FloorNumber){
                this.MoveDown(FloorNumber);
            }    
            else if (this.currentFloor < FloorNumber){
                this.MoveUp(FloorNumber);
            }    
            else {
                this.OpenDoor();
            }    
            
           
           Console.WriteLine("MoveNext")
        }
       static void MoveDown(int FloorNumber )
       {
           this.direction = 'down'; 
           this.status = 'Moving';
           var interval = setInterval(() => {
            this.currentFloor = this.currentFloor - 1;
            console.writeLine(this.currentFloor);
             if (this.currentFloor == FloorNumber) {
                 clearInterval(interval);
                console.writeLine("Arrived at floor " + this.currentFloor);
               this.OpenDoor();
             }
        }, 1000);
                    
           Console.WriteLine("MoveDown" + FloorNumber.ToString() )
       }
       public void MoveUp(int FloorNumber )
       this.direction = 'down'; 
           this.status = 'Moving';
           var interval = setInterval(() => {
            this.currentFloor = this.currentFloor + 1;
            console.writeLine(this.currentFloor);
             if (this.currentFloor == FloorNumber) {
                 clearInterval(interval);
                console.writeLine("Arrived at floor " + this.currentFloor);
               this.OpenDoor();
            }
        }, 1000);
                   

       {
           Console.WriteLine("MoveUp"+ FloorNumber.ToString)
       }
       public void addFloorToList(int FloorNumber)
       {
           Console.WriteLine("addFloorToList" + FloorNumber.ToString() );
       }
       public void OpenDoor()
       {
           Console.WriteLine("OpenDoor");
       }
       public void CloseDoor()
       {
           Console.WriteLine("CloseDoor");
       }
       public void activateInsideButton(int FloorNumber)
       {
           Console.WriteLine("activateInsideButton" + FloorNumber.ToString() );
       }
    }
}

File: Residential_Controller.js
class Column {
    constructor(number_of_floor, number_of_elevator){
      this.number_of_floor = number_of_floor;
      this.number_of_elevator = number_of_elevator;
      this.elevatorList = [];


      for (let i = 0; i < this.number_of_elevator; i++) {
             let elevator = new Elevator(i+1,number_of_floor)
             this.elevatorList.push(elevator);
        }
    }
}
class Button{
    constructor (direction, request_floor) {
        this.direction = direction;
        this.request_floor = request_floor;
        this.activate_button = false;
    }
}

class InsideButton{
    constructor (floor) {
        this.floor = floor;
        this.status = "desctivated";

    }
}
class ElevatorController {
    constructor(number_of_floor, number_of_elevator) {
      this.number_of_floor = number_of_floor;
      this.number_of_elevator = number_of_elevator;
      this.column = new Column(number_of_floor, number_of_elevator);
      this.button_list = [new Button()];

    }


    RequestElevator(floor_number, direction) {
        var selected_elevator = this.FindElevator(floor_number, direction);
        selected_elevator.addFloorToList(floor_number);
        selected_elevator.activateInsideButton(floor_number);
    }


    RequestFloor(elevator, floor_number) {
        elevator.activateInsideButton(floor_number);
        elevator.addFloorToList(floor_number);
        elevator.move_next();

    }
    FindElevator(FloorNumber) {
        var distance_floor = 999;
        var selected_elevator = null;
        for (var i = 0; i < this.column.elevatorList.length; i++) {
          var difference_floor = Math.abs(FloorNumber - this.column.elevatorList[i].current_floor);
            if (difference_floor < distance_floor) {
              distance_floor = difference_floor;
              selected_elevator =  this.column.elevatorList[i];
            }
           // IF Floor = Elevator.CurrentFloor AND Status = Stopped OR ElevatorDirection = Direction THEN
           //if (floor_number === this.column.elevatorList[i].CurrentFloor && this.column.elevatorList[i].status === "idle" && this.column.elevatorList[i].direction === direction) {
               //return this.elevatorList[i]
           //}
           //console.log(this.column.elevatorList[i].status)
       }
       return selected_elevator;
    }
}

class Elevator {
    constructor(elevator_number, number_of_floor) {
        this.elevator_number = elevator_number;
        this.direction = "NONE";
        this.status = "idle";
        this.floor_list = [];
        this.button_list = [];
        for (var i = 0; i < number_of_floor; i++) {
            this.button_list.push(new InsideButton(i));
        }
        this.current_floor = 1;
    }
    move_next(){
        let floor_list = this.floor_list
        let floor_number = floor_list.shift();
        if (this.current_floor > floor_number){
            this.moveDown(floor_number);
        }
        else if (this.current_floor < floor_number){
            this.moveUp(floor_number);
        }
        else {
            this.OpenDoor();
        }
    }
    moveDown(floor_number) {
        console.log ("Elevator is going down");
        this.direction = 'down';
        this.status = 'Moving';
        let interval = setInterval(() => {
            this.current_floor = this.current_floor - 1
            console.log(this.current_floor)
            if (this.current_floor == floor_number) {
                clearInterval(interval)
                console.log("Arrived at floor " + this.current_floor)
                this.OpenDoor()
            }
        }, 1000)
    }

    moveUp(floor_number) {
        console.log ("Elevator is going up");
        this.direction = 'up';
        this.status = 'Moving';
        let interval = setInterval(() => {
            this.current_floor = this.current_floor + 1
            console.log(this.current_floor)
            if (this.current_floor == floor_number) {
                clearInterval(interval)
                console.log("Arrived at floor " + this.current_floor)
                this.OpenDoor()
            }
        }, 1000)
    }
    addFloorToList(floor_number){
        this.floor_list.push(floor_number);
        if (this.direction == "Up"){
            this.floor_list.sort();
            console.log(this.floor_list)
        }
        else if (this.direction == "Down"){
         this.floor_list.sort().reverse();
           console.log(this.floor_list)
        }
    }

    OpenDoor(){
        console.log("Opening door on floor " + this.current_floor)
        this.status = 'open_door';
        setTimeout(() => {
          this.closeDoor()
        }, 5000);
    }

    closeDoor(){
        console.log("Closing door");
        this.status = 'close_door';
        if (this.floor_list.length > 0) {
            this.move_next()
        }
    }
    activateInsideButton(floor_number) {
        console.log ("Activated button at floor " + floor_number);
        if  (this.request_floor == this.floor_list){
            this.activate_InsideButton = false;
            }
        if (this.request_floor < this.floor_list){
            this.moveUp();
            }
        else if (this.request_floor > this.floor_list){
            this.moveDown();
            }
    }
    //}
}
const a = new ElevatorController(10,2);

//scenario 1
//a.column.elevatorList[0].current_floor = 1;
//a.column.elevatorList[1].current_floor = 5;
//a.RequestElevator(2, 'Up')
//a.RequestFloor(a.column.elevatorList[0], 6)


//scenario 2
//a.column.elevatorList[0].current_floor = 9;
//a.column.elevatorList[1].current_floor = 2;

//a.RequestElevator(0, 'Up')
//a.RequestFloor(a.column.elevatorList[1], 5)

//a.RequestElevator(2, 'Up')
//a.RequestFloor(a.column.elevatorList[1], 4)

//a.RequestElevator(8, 'Down')
//a.RequestFloor(a.column.elevatorList[0], 1)

//Scenario 3
// a.column.elevatorList[0].current_floor = 9;
// a.column.elevatorList[1].current_floor = 2;
//a.RequestElevator(9, 'Down')
//a.RequestFloor(a.column.elevatorList[0], 2)
//a.RequestElevator(2, 'Down')
//a.RequestFloor(a.column.elevatorList[1], 1)






